### ISSUE TITLE
SetNillable with nil value

### ISSUE THREAD

**Issue by erlendp** [ent/ent#2108]
Raising this as a new issue, as I feel this should probably be addressed.

The behaviour of `SetNillableFieldX` functions, when passing in a nil value, appears to be to do nothing at all.

I would expect a function for setting a nillable field to set the field to nil when a nil value is passed in. Anything else seems surprising to me.

I understand that this can be worked around by defining a custom template, but I'd rather see this addressed in the core API to prevent confusing to users.

Is there any compelling reason for not non-standard behaviour?

**Comment by godcong** [comment]
nillable field has ClearXXX functions, it can clear the setted values

**Comment by erlendp** [comment]
Yes, I'm aware of the ClearX functions, but why the decision to add a separate function for this, when the SetNillableX functions could have served this purpose, seemingly in a more intuitive way?

As it is currently, this design will lead to a lot of code like this:

`if field == nil {
    entity.ClearField()
} else {
    entity.SetNillableField(field)
`

which could be simplified to just:

`entity.SetNillableField(field)`

It also raises the question of why even have the SetNillableX functions at all. If they only ever have any effect with non-nil values, you could just have the SetX functions, and dereference once you know your value is non-nil.

**Comment by a8m** [comment]
Hey @erlendp 👋
Thanks for raising up this question and thanks @godcong for the support. It's not the first time this question is asked here, so I think it's worth adding it to our website.

The `SetNillable<F>` modifier was added to the `<T>Update` builders to simplify the process of converting API inputs (e.g. GraphQL mutation input or REST PATCH input) to modifier calls. For example, suppose you have a `User` entity with 2 mutable fields, "name" and "age", and you want to update them through an API endpoint. Now, since the 2 fields in the API input are optional (pointers), your handler code will look as follows:

```go
u := client.User.UpdateOne(id)
if input.Age != nil {
	u.SetAge(input.Age)
}
if input.Name != nil {
	u.SetName(input.Name)
}
```

And if you have more fields, this block becomes too long with repetitive checks:

```go
u := client.User.UpdateOne(id)
if input.Age != nil {
	u.SetAge(input.Age)
}
if input.Name != nil {
	u.SetName(input.Name)
}
// ....
if input.Address != nil {
	u.SetAddress(input.Address)
}
```

So, the idea was to provide an API that will reduce (for us) all these repetitive checks, and it's called `SetNillable<F>`. And as documented in the godoc, `SetNillable<F>(v)` updates `F` if `v` is not nil. 

```go
u := client.User.UpdateOne(id).
	SetNillableAge(input.Age).
	SetNillableName(input.Name).
	// ...
	SetNillableAddress(input.Address)
```

For cases like you mentioned: "Set or clear" the fields, I suggest using a custom template to generate a function like `SetOrClear<F>` optional fields. See my comment in https://github.com/ent/ent/issues/1119#issuecomment-753588425:

> Created you some example to start working with it.
> 
> ```go
> {{ define "nullsetter" }}
> 	{{/* Add the base header for the generated file */}}
> 	{{ $pkg := base $.Config.Package }}
> 	{{ template "header" $ }}
> 
> 	{{/* Loop over all updaters and implement the "SetOrClear" method for all optional fields */}}
> 	{{ range $n := $.Nodes }}
> 		{{ $updater := $n.UpdateName }}
> 		{{ range $f := $n.Fields }}
> 			{{ if $f.Optional }}
> 				{{ $set := print "Set" $f.StructField }}
> 				{{ $clear := print "Clear" $f.StructField }}
> 				func (u *{{ $updater }}) SetOrClear{{ $f.StructField }}(value *{{ $f.Type }}) *{{ $updater }} {
> 					if value == nil {
> 						return u.{{ $clear }}()
> 					}
> 					return u.{{ $set }}(*value)
> 				}
> 			{{ end }}
> 		{{ end }}
> 	{{ end }}
> {{ end }}
> ```

Please, let me know if it does answer your question, you need some help, or you want to contribute to ent and add this information to our website. Thanks 🙏 😃

**Comment by erlendp** [comment]
Thanks for your explanation - I had a feeling this would be the answer. I believe the crux of the problem is in the semantics of what nil represents. In your use case, nil represents a field that should be ignored, and in my use case it represents a field that should actually be cleared. Given that this decision relates to the client API semantics, I personally feel it should be a decision made by the client, and not by the ORM library. In a pure ORM library, nil should represent the value you intend the set in the database, rather than trying to predict what semantics the user had intended. Then a custom template could have been specified for your use case, or a different function named appropriately (`SetIfNonNilX`?) if the use case is common enough.

All that said - I understand that this would be a fairly big breaking change at this point, and is unlikely to be changed, but I do think this should be called out specifically in the documentation to avoid this confusion in future, and potentially a new function added to cater to the nil == clear use case, if possible.

**Comment by theterminalguy** [comment]
@a8m once caveat with the `SetNillableX` method though is that it seems to be only generated for Optional fields. If the field is required, then there's no method currently in the core API that lets you SetNillable for the required field, at least not that I am aware of. 

Case in point, in a REST API, let's say all fields all required for a POST request and I want my users to do a PUT/PATCH by providing only the fields they'd like updated. Currently, you'd still have to write a ton of code to get this to work except those fields are marked optional when defining your schema.

**Comment by a8m** [comment]
> @a8m once caveat with the `SetNillableX` method though is that it seems to be only generated for Optional fields. If the field is required, then there's no method currently in the core API that lets you SetNillable for the required field, at least not that I am aware of.
> 
> Case in point, in a REST API, let's say all fields all required for a POST request and I want my users to do a PUT/PATCH by providing only the fields they'd like updated. Currently, you'd still have to write a ton of code to get this to work except those fields are marked optional when defining your schema.

Valid point. I can handle it in this release.

**Comment by endigma** [comment]
Any updates on this?

**Comment by waltcow** [comment]
Any updates on this?

**Comment by theterminalguy** [comment]
If anyone is looking for a potential workaround with regards [to my use case](https://github.com/ent/ent/issues/2108#issuecomment-961898661), you can use an [ent builder](https://github.com/ent/ent/blob/master/dialect/sql/builder.go) to achieve this. Here is a **contrived example** :

```go
// setNillableStringField is a helper function that calls the callback function only if the value is not empty
func setNillableStringField(val string, cb func(v string) error) error {
	if val != "" {
		err := cb(val)
		if err != nil {
			return err
		}
	}
	return nil
}

type UserParams struct {
	FirstName string `json:"first_name" validate:"required,min=2,max=50"`
}

type UserRepository struct {}

func (r *UserRepository) Update(id string, params *UserParams) error {
	record, _ := r.Find(id) // ignore error for brevity
	bldr := record.Update() // returns a builder

	var vldErrs []error // collect all validation errors

	if vldErr := setNillableStringField(record.FirstName, func(v string) error {
		// validate.StructPartial is provided by go validator
		// feel free to provide your own validation function or do whatever you want
		if err := validate.StructPartial(params, "FirstName"); err != nil {
			return err
		}
		bldr.SetFirstName(v)
		return nil
	}; vldErr != nil {
		vldErrs = append(vldErrs, vldErr)
	}
	bld.Save(ctx) // this commit the changes to the database
	if len(vldErrs) > 0 {
		return fmt.Errorf("validation errors: %v", vldErrs)
	}
	return nil
}
```
In summary, what we did is use a builder to **only update the fields that are provided in the request**. "Provided" in this case means that the field(s) are not empty. If the field is empty, we do not update it, hence no validation error is returned for fields that are required at the database level. 

The setNillableStringField function does the trick of only calling the callback function if the value is NOT empty. The builder is then used to set the field that needs to be updated. 

You can, however, re-write the `setNillableStringField` with [generic types](https://go.dev/doc/tutorial/generics) and use it for all types of fields. This will make the code more DRY and avoid things like SetNillableStringField, SetNillableBoolField, SetNillableIntField, etc.

**Comment by theterminalguy** [comment]
Cool, I'll give this a try. Thanks @endigma
